src/sync/sync_manager.hpp

```diff
@@ -155,8 +156,30 @@ friend class SyncSession;
     // Protects m_users
     mutable std::mutex m_user_mutex;
 
-    // A map of user identities to (shared pointers to) SyncUser objects.
-    std::unordered_map<std::string, std::shared_ptr<SyncUser>> m_users;
+    // A custom key for the user map system. It's basically a hashable tuple.
+    struct UserMapKey {
+        std::string user_id;
+        std::string auth_server_url;
+
+        bool operator==(const UserMapKey& other) const
+        {
+            return user_id == other.user_id && auth_server_url == other.auth_server_url;
+        }
+
+        struct Hasher {
```

Comment:
I'd suggest specializing `std::hash` instead, so the type works as-is without requiring the hash function be explicitly specified when using the type with `std::unordered_map`. You can see list.hpp for an example of how we do this for `List`.